### ISSUE TITLE
[BUG] arrayfire-cpu3-mkl and arrayfire-opencl3-mkl are broken on Ubuntu 20.04

### ISSUE THREAD

**Issue by BA8F0D39** [arrayfire/arrayfire#3116]
Description
===========
Installing arrayfire, arrayfire-cpu3-mkl, arrayfire-opencl3-mkl, and arrayfire-unified3 via APT on Ubuntu 20.04 results in a broken arrayfire installation. Arrayfire can't find the unified backend.

arrayfire, arrayfire-cpu3-openblas, arrayfire-opencl3-openblas, and arrayfire-unified3 via APT works but doesn't have multithreading on CPU backend.

**Comment by 9prady9** [comment]
@BA8F0D39 Can you please share the exact commands you used to install arrayfire using apt.

**Comment by BA8F0D39** [comment]
docker run --name arrayfire --gpus all   -it  nvidia/cuda:11.2.2-cudnn8-runtime-ubuntu20.04   bash
apt update
apt install gnupg2 ca-certificates apt-utils software-properties-common wget  build-essential git cmake libfreeimage-dev  cmake-curses-gui  
cd /tmp
wget https://apt.repos.intel.com/intel-gpg-keys/GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB
apt-key add GPG-PUB-KEY-INTEL-SW-PRODUCTS-2019.PUB
sh -c 'echo deb https://apt.repos.intel.com/mkl all main > /etc/apt/sources.list.d/intel-mkl.list'
apt update
apt-key adv --fetch-key https://repo.arrayfire.com/GPG-PUB-KEY-ARRAYFIRE-2020.PUB
echo "deb [arch=amd64] https://repo.arrayfire.com/ubuntu focal main" | tee /etc/apt/sources.list.d/arrayfire.list
apt update
apt install arrayfire arrayfire-cuda3-cuda-11-1  arrayfire-cpu3-mkl  arrayfire-opencl3-mkl  




debconf: unable to initialize frontend: Dialog
debconf: (No usable dialog-like program is installed, so the dialog based frontend cannot be used. at /usr/share/perl5/Debconf/FrontEnd/Dialog.pm line 76, <> line 10.)
debconf: falling back to frontend: Readline
Selecting previously unselected package arrayfire-cpu3-mkl.
(Reading database ... 23428 files and directories currently installed.)
Preparing to unpack .../0-arrayfire-cpu3-mkl_3.8.0_amd64.deb ...
Unpacking arrayfire-cpu3-mkl (3.8.0) ...
Selecting previously unselected package arrayfire-opencl3-mkl.
Preparing to unpack .../1-arrayfire-opencl3-mkl_3.8.0_amd64.deb ...
Unpacking arrayfire-opencl3-mkl (3.8.0) ...
Selecting previously unselected package arrayfire-cuda3-cuda-11-1.
Preparing to unpack .../2-arrayfire-cuda3-cuda-11-1_3.8.0_amd64.deb ...
Unpacking arrayfire-cuda3-cuda-11-1 (3.8.0) ...
Selecting previously unselected package arrayfire-unified3.
Preparing to unpack .../3-arrayfire-unified3_3.8.0_amd64.deb ...
Unpacking arrayfire-unified3 (3.8.0) ...
Selecting previously unselected package arrayfire-headers.
Preparing to unpack .../4-arrayfire-headers_3.8.0_amd64.deb ...
Unpacking arrayfire-headers (3.8.0) ...
Selecting previously unselected package arrayfire-cmake.
Preparing to unpack .../5-arrayfire-cmake_3.8.0_amd64.deb ...
Unpacking arrayfire-cmake (3.8.0) ...
Selecting previously unselected package arrayfire-examples.
Preparing to unpack .../6-arrayfire-examples_3.8.0_amd64.deb ...
Unpacking arrayfire-examples (3.8.0) ...
Selecting previously unselected package arrayfire-doc.
Preparing to unpack .../7-arrayfire-doc_3.8.0_amd64.deb ...
Unpacking arrayfire-doc (3.8.0) ...
Selecting previously unselected package arrayfire.
Preparing to unpack .../8-arrayfire_3.8.0_amd64.deb ...
Unpacking arrayfire (3.8.0) ...
Preparing to unpack .../9-libnvidia-compute-460-server_460.32.03-0ubuntu0.20.04.1_amd64.deb ...
Unpacking libnvidia-compute-460-server:amd64 (460.32.03-0ubuntu0.20.04.1) ...
dpkg: error processing archive /tmp/apt-dpkg-install-U16xpj/9-libnvidia-compute-460-server_460.32.03-0ubuntu0.20.04.1_amd64.deb (--unpack):
 unable to make backup link of './usr/lib/x86_64-linux-gnu/libcuda.so.460.32.03' before installing new version: Invalid cross-device link
dpkg-deb: error: paste subprocess was killed by signal (Broken pipe)
Errors were encountered while processing:
 /tmp/apt-dpkg-install-U16xpj/9-libnvidia-compute-460-server_460.32.03-0ubuntu0.20.04.1_amd64.deb
E: Sub-process /usr/bin/dpkg returned an error code (1)

**Comment by 9prady9** [comment]
@BA8F0D39 There is one difference what your setup does and what I tried and I was able to install arrayfire successfully.

```
Unpacking arrayfire-doc (3.8.0) ...
Selecting previously unselected package arrayfire.
Preparing to unpack .../18-arrayfire_3.8.0_amd64.deb ...
Unpacking arrayfire (3.8.0) ...
Selecting previously unselected package libnvidia-compute-460-server:amd64.
Preparing to unpack .../19-libnvidia-compute-460-server_460.32.03-0ubuntu0.20.04.1_amd64.deb ...
Unpacking libnvidia-compute-460-server:amd64 (460.32.03-0ubuntu0.20.04.1) ...
Setting up arrayfire-cmake (3.8.0) ...
Setting up arrayfire-examples (3.8.0) ...
Setting up arrayfire-headers (3.8.0) ...
Setting up intel-openmp-19.1.0-166 (19.1.0-166) ...
Setting up intel-comp-l-all-vars-19.1.0-166 (19.1.0-166) ...
Setting up intel-mkl-doc-2020 (2020.4-304) ...
Setting up intel-tbb-libs-2020.0-166 (2020.0-166) ...
Setting up intel-comp-nomcu-vars-19.1.0-166 (19.1.0-166) ...
Setting up ocl-icd-libopencl1:amd64 (2.2.11-1ubuntu1) ...
Setting up cuda-nvrtc-11-1 (11.1.105-1) ...
Setting up arrayfire-doc (3.8.0) ...
Setting up libnvidia-compute-460-server:amd64 (460.32.03-0ubuntu0.20.04.1) ...
Setting up arrayfire-cuda3-cuda-11-1 (3.8.0) ...
Setting up arrayfire-unified3 (3.8.0) ...
Setting up intel-mkl-common-2020.0-166 (2020.0-166) ...
Setting up intel-mkl-core-rt-2020.0-166 (2020.0-166) ...
Setting up intel-mkl-gnu-rt-2020.0-166 (2020.0-166) ...
Setting up arrayfire-cpu3-mkl (3.8.0) ...
Setting up arrayfire-opencl3-mkl (3.8.0) ...
Setting up arrayfire (3.8.0) ...
Processing triggers for libc-bin (2.31-0ubuntu9.2) ...
```
The `--gpus all` argument which brings in host driver to container I think and the error message also kind points to the same - failing to create cuda driver link or something like that. Is this the problem you are facing or something else ? Because the original description says you were having issues with using unified backend but not installation. I can confirm that using the same steps you shared I was able to finish installation and I can find the unified library also under `/usr/lib`

> but doesn't have multithreading on CPU backend.

Can you elaborate on what do you mean by multi-threading here ?

**Comment by BA8F0D39** [comment]
@9prady9 
I removed --gpus all in docker and I can install arrayfire successfully. However, the linker can't detect Intel MLK libraries for some reason.

[0] Intel: Intel(R) Core(TM) i5-8400 CPU @ 2.80GHzINTEL MKL ERROR: /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64/libmkl_avx2.so: undefined symbol: mkl_sparse_optimize_bsr_trsm_i8.
Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.

**Comment by 9prady9** [comment]
@BA8F0D39 I am not sure clearly why docker containers fail to use nvidia runtime and I have been facing this issue very recently irrespective of the base image I am using. My two cents; recent driver upgrade started causing this. I have 460.67 driver. Which driver do you have on the host ?

Have you setup MKL before using it ? Did you do the following before using MKL
```
source /opt/intel/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64
```

**Comment by BA8F0D39** [comment]
@9prady9 
I used
`source /opt/intel/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64`

However, the error is still the same.
```
INTEL MKL ERROR: /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_avx2.so: undefined symbol: mkl_sparse_optimize_bsr_trsm_i8.
Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.
```

**Comment by 9prady9** [comment]
@BA8F0D39 Can you share the cmake or makefile or command that you are using to compile your code. Something like a standalone snippet that I can run in this docker image to reproduce the problem.

**Comment by BA8F0D39** [comment]
@9prady9 

Sparse matrix operations have linking errors in the rustc compiler.



```

	arrayfire::set_backend(arrayfire::Backend::CPU);
	let dims = arrayfire::Dim4::new(&[4,4,1,1]);
	let a = arrayfire::randu::<f32>(dims);
	let sa =  arrayfire::sparse_from_dense(&a ,arrayfire::SparseFormat::CSR);
	let b = arrayfire::randu::<f32>(dims);
	let sb = arrayfire::sparse_from_dense(&b ,arrayfire::SparseFormat::CSR);
	let c =  arrayfire::matmul(&a, &b, arrayfire::MatProp::NONE, arrayfire::MatProp::NONE);
	
	arrayfire::af_print!("Random matrix", c);

```



Name: Intel
Platform: CPU
Toolkit: GNU Compiler Collection(GCC/G++) 7.5.0
Compute: 0.0
Revision: d99887ae
[0] Intel: Intel(R) Core(TM) i5-8400 CPU @ 2.80GHzRandom matrix
[4 4 1 1]
INTEL MKL ERROR: /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64_lin/libmkl_avx2.so: undefined symbol: mkl_sparse_optimize_bsr_trsm_i8.
Intel MKL FATAL ERROR: Cannot load libmkl_avx2.so or libmkl_def.so.

**Comment by 9prady9** [comment]
```
source /opt/intel/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64
```
this alone is sufficient instead of the doing all the following

```
source /opt/intel/compilers_and_libraries_2020.0.166/linux/bin/compilervars.sh intel64
source /opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/bin/mklvars.sh intel64

export LD_LIBRARY_PATH=/opt/intel/compilers_and_libraries_2020.0.166/linux/compiler/lib/intel64/
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/intel/compilers_and_libraries_2020.0.166/linux/mkl/lib/intel64/
export LD_LIBRARY_PATH=$LD_LIBRARY_PATH:/opt/intel/compilers_and_libraries_2020.0.166/linux/tbb/lib/intel64/gcc4.8/
```

vola!, I was able to reproduce the exact issue on my end inside the docker image I created based on your instructions. This is unfortunate bug, I will look into and let you know once I have an update.

**Comment by 9prady9** [comment]
Interesting, when running the rust program only it is happening. C/C++ examples work fine. @BA8F0D39 Is that the case for you as well ?

**Comment by BA8F0D39** [comment]
@9prady9 
It might be the sparse functions behaving differently in Intel MKL compared to different backend like openBLAS and CUDA?

**Comment by 9prady9** [comment]
@BA8F0D39 No, that is not the cause. We have already identified the problem it has to do with runtime loading of MKL related libraries. We are already working on a fix and will update we fix the problem.

**Comment by 9prady9** [comment]
A fix for this has been pushed in PR #3120 https://github.com/arrayfire/arrayfire/pull/3120/commits/886c92f1712da6afffcdc6d8db0ce62b471a0ceb

@BA8F0D39 We will post another update once we update our packages.
